README.md
# Lynda_Downloader
<img src="https://i.ibb.co/B3CjZ3x/4ffb2ec696f5da0c09fbca3a1a0f3753.png"></img>

Lynda Downloader is a python script that uses Selenium Driver and is capable to download multiple courses at once, their exercise files and related information from Lynda.com. 

The idea of this scrtipt was to be able to download multiple courses with a given list along with their excersize files and information about each course and then to display them directly on a private library website, without the need of touching anything else. It also dump the video duration from the download link and concat the information in the video name, so you can extract it direclty without the need to use additional calculations.

The script will create categgory and course name folders automatically and download the video files inside along with log.txt file in which all downloaded videos paths will be recorder. After the download completion you can move all .zip files to the related course. That is helpfull if you want to create a private library or to build similar to Lynda.com website.</br>

In suppport for this script I have developed a <a href="https://github.com/r00tmebaby/LyndaWebSite"> LyndaWeb</a>  where all downloaded videos can be displayed and accessed easily.


# Requirements
1. Python 3+ installed </br>
2. Selenium driver installed. </br>
3. PIP Installed
4. Colorama Library Installed
5. TQDM Library Installed
6. Chrome browser installed and chromedriver downloaded</br>

# Installation
1. You need to download and install python version 3 or higher</br>
2. You can then install selenium driver by typing <code><b>pip install selenium</b></code> in the command prompt</br>
2. You can then install tqdm by typing <code><b>pip install tqdm</b></code> in the command prompt</br>
2. You can then install colorama by typing <code><b>pip install colorama</b></code> in the command prompt</br>
3. Now is the time to install Chrome Browser and download Chrome driver


# Configuration
1. Specify your login credentials in config.py, login url, courses list file destination, chromedriver destination and main download folder
*Important note:
    From the last update if you are using your chrome browser path and have logged to lynda at leeast once, then the session is still valid and you can start lynda Downloader and it will login automatically without the need to fill the credentials in the config file.
2. You have to create a donwload list manualy with the links to the courses that you want to download
3. Run the script by typing lindi.py and it will do the work for you.



CHANGELOG.md
# Change Log

## 1.1 (2019-31-01)
Bugfixes:
  - Fixed "Can not get video url" message. Slow internet connection needs more time to get the video links
  - Fixed modal popup message "Course Retiring Soon" that shows up on some courses  

## 1.0 (2019-29-01)

Features:
  - Added an option to turn off the downloading progress bar. 
  It slowing down the downloading speed significantly and in longer course lists is better to be off
  - Added an option to delete all succesfully downloaded links from the list.
  - Added courses dump function at the bottom of functions.py file.
  Useful if you want to download all newely added courses at some point

Bugfixes:
  - Fixed "an empty url message"




Config.py
'''

Author 	: Zdravko Georgiev (r00tmebaby)
Github 	: https://github.com/r00tmebaby
License : MIT


Copyright (c) 2019 Zdravko Georgiev (r00tmebaby)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the
Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'''


##############################################
# Lynda Downloading Script
#   by r00tme 26/01/2019
# version: 0.7 Downloading with Pre-Specified List
###############################################
#               Options:
# Creating Category Folder Automatically
# Creating Course Sub-folder Automatically
# Creating Course Difficulty Folder Prefix Automatically
# Creating .php File with a Course Description Automatically - Useful for web development
# Dumping video duration and date added
# Creating a log.txt file to  save the course path, so the .zip archives can be moved later
#############################################

path = r"C:\\chromedriver.exe"  # Specify chrome Driver Path @String
browser = 1  # 1= Chrome, 2= Firefox, 3=Safari, 4=Opera :: Any Other Number is Default Chrome
# Replace r00tme with your username, the script will try to use your Chrome settings
profile_dr = r"C:\\Users\\r00tme\\AppData\\Local\\Google\\Chrome\\User Data\\Default"
page = ""
maxi = False  # Maximize Window @Boolean
load = 30  # PageLoad Time @Integer

save_dir = r"Z:\\F\\Tutorials\\"  # Saving Files Main Directory
finished_list = r"exercise_logs.txt"  # File that will log all downloaded course directories so the .zip archives can be moved later
course_list = r"Z:\\F\\dls_list.txt"  # All courses that we want to be downloaded. Courses list
zip_download = r"Z:\\F\\Download\\"  # Directory that will be used to download zip archives, they will be moved to their folders at the end
show_progress = False

####--- Login Type--#####
login_in = True  # Login in
username = ""  # Login Username
password = ""  # Login Password
usr_field = "username"  # Field Name For Username :: Selected By Name
pas_field = "password"  # Field Name For Password :: Selected By Name
login_btn = "btn-primary"  # Login Button :: Selected By Class Name
#########################


Functions.py
'''

Author 	: Zdravko Georgiev (r00tmebaby)
Github 	: https://github.com/r00tmebaby
License : MIT


Copyright (c) 2019 Zdravko Georgiev (r00tmebaby)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the
Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'''


##############################################
# Lynda Downloading Script
#   by r00tme 26/01/2019
# version: 1.0 Downloading with Pre-Specified List
###############################################
#               Options:
# Creating Category Folder Automatically
# Creating Course Sub-folder Automatically
# Creating Course Difficulty Folder Prefix Automatically
# Creating .php File with a Course Description Automatically - Useful for web development
# Dumping video duration and date added
# Creating a log.txt file to  save the course path, so the .zip archives can be moved later
#############################################


from selenium import webdriver
from pathlib import Path
from tqdm import tqdm
from colorama import init, Fore, Back, Style
import urllib.request, urllib.request, ssl, urllib3, Config, os

class Main:

    # Method Driver Return Driver Object
    # @param  browser_type Integer -> Switch between Different Browsers 1-4
    # @param  driver_path  String -> Path to the driver
    ##################################################
    def driver(self, browser_type=1, driver_path='C:/chromedriver.exe'):
        if browser_type == 1:
            driver = webdriver.Chrome(driver_path, chrome_options=self.build_chrome_options())
        elif browser_type == 2:
            driver = webdriver.Firefox(driver_path)
        elif browser_type == 3:
            driver = webdriver.Safari(driver_path)
        elif browser_type == 4:
            driver = webdriver.Opera(driver_path)
        else:
            driver = webdriver.Chrome(driver_path)
        return driver

    def build_chrome_options (self, profile_dir = Config.profile_dr):

        chrome_options = webdriver.ChromeOptions()
        chrome_options.accept_untrusted_certs = True
        chrome_options.assume_untrusted_cert_issuer = True
        # chrome configuration
        # More: https://github.com/SeleniumHQ/docker-selenium/issues/89
        # And: https://github.com/SeleniumHQ/docker-selenium/issues/87
        chrome_options.add_argument("--disable-extensions")
        chrome_options.add_argument("--no-sandbox")
        chrome_options.add_argument("--disable-gpu")
        chrome_options.add_argument(
            '--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36')
        chrome_options.add_argument("--disable-impl-side-painting")
        chrome_options.add_argument("--user-data-dir=" + os.path.abspath(profile_dir))
        chrome_options.add_argument("--disable-setuid-sandbox")
        chrome_options.add_argument("--disable-seccomp-filter-sandbox")
        chrome_options.add_argument("--disable-breakpad")
        chrome_options.add_argument("--disable-client-side-phishing-detection")
        chrome_options.add_argument("--disable-cast")
        chrome_options.add_argument("--disable-cast-streaming-hw-encoding")
        chrome_options.add_argument("--disable-cloud-import")
        chrome_options.add_argument("--disable-popup-blocking")
        chrome_options.add_argument("--ignore-certificate-errors")
        chrome_options.add_argument("--disable-session-crashed-bubble")
        chrome_options.add_argument("--disable-ipv6")
        chrome_options.add_argument("--allow-http-screen-capture")
        return chrome_options
        ##################################################

    # Static Method Download Download File
    ###################################################
    @staticmethod
    def download(url, names):
        urllib3.disable_warnings()

        try:
            _create_unverified_https_context = ssl._create_unverified_context
        except AttributeError:
            pass
        else:
            ssl._create_default_https_context = _create_unverified_https_context

        if urllib.request.urlretrieve(url, names):
            return True
        else:
            return False

    ##################################################
    #  Progress bar Class tqdm
    ###################################################
    class progress_bar(tqdm):
        def update_to(self, b=1, bsize=1, tsize=None):
            if tsize is not None:
                self.total = tsize
            self.update(b * bsize - self.n)

    def show_progress_bar(self, url, filename, output_path):
        with self.progress_bar(unit='B',smoothing=0.3, unit_scale=True,
                                 miniters=1, desc=filename) as t:
            urllib.request.urlretrieve(url, filename=output_path, reporthook=t.update_to)

    ##################################################
    #  Background and Font Colours
    # Reference => *Nasir Khan (r0ot h3x49)  - **https://github.com/r0oth3x49** ##*
    ###################################################
    class bcolors:
        if os.name == "posix":
            init(autoreset=True)
            # colors foreground text:
            fc = "\033[0;96m"
            fg = "\033[0;92m"
            fw = "\033[0;97m"
            fr = "\033[0;91m"
            fb = "\033[0;94m"
            fy = "\033[0;33m"
            fm = "\033[0;35m"

            # colors background text:
            bc = "\033[46m"
            bg = "\033[42m"
            bw = "\033[47m"
            br = "\033[41m"
            bb = "\033[44m"
            by = "\033[43m"
            bm = "\033[45m"

            # colors style text:
            sd = Style.DIM
            sn = Style.NORMAL
            sb = Style.BRIGHT
        else:
            init(autoreset=True)
            # colors foreground text:
            fc = Fore.CYAN
            fg = Fore.GREEN
            fw = Fore.WHITE
            fr = Fore.RED
            fb = Fore.BLUE
            fy = Fore.YELLOW
            fm = Fore.MAGENTA

            # colors background text:
            bc = Back.CYAN
            bg = Back.GREEN
            bw = Back.WHITE
            br = Back.RED
            bb = Back.BLUE
            by = Fore.YELLOW
            bm = Fore.MAGENTA

            # colors style text:
            sd = Style.DIM
            sn = Style.NORMAL
            sb = Style.BRIGHT

    ##################################################
    #  Method Waits wait time
    # @param  wait Integer
    ###################################################

    def waits(self, wait):
        self.driver().implicitly_wait(wait)

    ##################################################
    #  Method Maximizing Window
    # @param  m_window Boolean -> True by Default
    ###################################################

    def max_window(self, m_window=True):
        if m_window:
            self.driver().maximize_window()

    ##################################################
    #  Method Looking for file existence before downloading
    # @param  m_window Boolean -> True by Default
    ###################################################
    def file_exist(self, path):

        if Path(path).is_file():
            return True
        else:
            return False

    ##################################################
    #  Method checking whether the directory exists or not
    # @param  Directory Path String
    ###################################################
    def dir_exist(self, path):
        if Path(path).is_dir():
            return True
        else:
            return False


Lindi.py
'''

Author 	: Zdravko Georgiev (r00tmebaby)
Github 	: https://github.com/r00tmebaby
License : MIT


Copyright (c) 2019 Zdravko Georgiev (r00tmebaby)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the
Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'''


##############################################
# Lynda Downloading Script
#   by r00tme 26/01/2019
# version: 1.0 Downloading with Pre-Specified List
###############################################
#               Options:
# Creating Category Folder Automatically
# Creating Course Sub-folder Automatically
# Creating Course Difficulty Folder Prefix Automatically
# Creating .php File with a Course Description Automatically - Useful for web development
# Dumping video duration and date added
# Creating a log.txt file to  save the course path, so the .zip archives can be moved later
#############################################

from Functions import Main
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import os, time, re, Config, sys, threading, itertools, datetime

main = Main()  # Calling Main Class
sys.stdout.write(
 "\r%s%s###############################################\n"
     "#     Lynda.com Downloading Script            #\n"
     "#     @author r00tmebaby  26/01/2019          #\n"
     "#     @version: 1.0                           #\n"
     "#     Downloading with Pre-Specified List     #\n"
     "##############################################\n\n" % (
     main.bcolors.sd, main.bcolors.fc))
sys.stdout.flush()
driver = main.driver(Config.browser, Config.path)  # The Driver
done = False

if not main.file_exist(Config.course_list):
    exit("\r%s[%s*%s] ERROR: %sFile %s does not exist in the specified directory. Please check the configuration file!" % (
        main.bcolors.bm, main.bcolors.fc, main.bcolors.fm, main.bcolors.fr, Config.course_list))

# Animation function
def animate():
    for c in itertools.cycle(['|', '/', '-', '\\']):
        if done:
            sys.stdout.write('\r%s[%s*%s]%s Processing URL: %s' %(
              main.bcolors.bm, main.bcolors.fc, main.bcolors.fm, main.bcolors.fc, driver.current_url))
            sys.stdout.flush()
            break
        sys.stdout.write('\r%s[%s%s%s]%s Loading Lynda.com and signing ' %(
            main.bcolors.bm, main.bcolors.fc, main.bcolors.sb+main.bcolors.fw+c, main.bcolors.sn+main.bcolors.fm, main.bcolors.fc))
        sys.stdout.flush()
        time.sleep(0.1)


threading.Thread(target=animate).start()   # Start threading process to the animate function
driver.set_page_load_timeout(Config.load)  # Using Load Time
main.max_window(Config.maxi)               # Maximize Window
driver.get(Config.page)                    # Starting Url

courses_count = 0       # Courses Counter
total_counter = 0       # Counting the total videos downloaded
prefix = ""             # This prefix will be put on front of every course to mark its difficulty level

# Login attempt, Please replace find_element_by_name if your login url needs different selection

if Config.login_in:
            driver.find_element_by_name(Config.usr_field).send_keys(Config.username)
            driver.find_element_by_name(Config.pas_field).send_keys(Config.password)
            driver.find_element_by_class_name(Config.login_btn).click()


done = True  # Resetting the loading animation

# Opening the course downloading list and reading the urls-s if they exit
for items_for_download in open(Config.course_list, "r"):
    normal_list = items_for_download.split("?")
    driver.get(normal_list[0])
   
    # Searching for a modal message and trying to close it
    try:
       driver.find_element_by_class_name("close").click()
       driver.find_element_by_class_name("close").click()
       time.sleep(0.2)
    except:
        pass
     
    cat_name = items_for_download.split("/")[3]
    if len(cat_name) == 0:
        cat_name =  items_for_download.split("/")[4]


    # Opening Overview TAB To inspect the Content - It is bit buggy sometimes and need couple of clicks
    driver.find_element_by_css_selector('.tab-overview.ga').click()
    driver.find_element_by_css_selector('.tab-overview.ga').click()
    driver.find_element_by_css_selector('.tab-overview.ga').click()
    driver.find_element_by_css_selector('.tab-overview.ga').click()

    # Selecting Course Difficulty Text
    elementss = driver.find_element_by_css_selector(
        "#course-info > div > div.col-sm-3.col-md-3.col-lg-2 > div:nth-child(1) > h6 > strong").text

    # Setting Course Name Prefix to mark the course difficulty - Useful for further developments
    if elementss == "Intermediate":
        prefix = "1-"
    elif elementss == "Advanced":
        prefix = "2-"
    else:
        prefix = "0-"

    # Stripping the course name from the course link
    if not items_for_download.split("/")[4] == cat_name:
        course_name = prefix + items_for_download.split("/")[4].replace("-", " ")
    else:
        course_name = prefix + items_for_download.split("/")[5].replace("-", " ")

    logs1 = open(Config.save_dir + "logs.txt", "a+")
    logs1.write("%s" % ("\n"+cat_name + r"\\" + course_name))

    # Creating Course directory if it does not exist
    if not main.dir_exist(Config.save_dir + r"\\" + cat_name + r"\\" + course_name):
        os.makedirs(Config.save_dir + r"\\" + cat_name + r"\\" + course_name)
        sys.stdout.write('\n%s[%s*%s]%s Directory %s has been successfully created' % (
            main.bcolors.bm, main.bcolors.fc, main.bcolors.fm, main.bcolors.fc,  cat_name +"\\" + course_name))
        sys.stdout.flush()
    try:
          driver.find_element_by_link_text('Exercise Files').click()
          # Trying to find exercise file if not found just continue to the next operation
          driver.find_element_by_link_text('Exercise Files').click()

          driver.find_element_by_link_text('Exercise Files').click()
          # downloading the exercise files
          sys.stdout.write('\n%s[%s*%s]%s Downloading Exercise Files.... ' % (
              main.bcolors.bm, main.bcolors.fc, main.bcolors.fm, main.bcolors.fc))
          sys.stdout.flush()
          ex_down_link = driver.find_element_by_css_selector('.course-file.clearfix.ga').click()
          driver.find_element_by_link_text('Exercise Files').click()
          driver.find_element_by_link_text('Exercise Files').click()
          tex_down_link = driver.find_element_by_partial_link_text('Ex_Files')
          time.sleep(0.2)
          logs = open(Config.save_dir + Config.finished_list, "a+")
          clean_f_name = re.sub(r'\([^()]*\)', '', tex_down_link.text).replace("_", "")
          logs.write("%s" % (cat_name + r"\\" + course_name + ":" + clean_f_name))

    except:
        driver.get(driver.current_url)  # Refresh the page if excercise files are not found

    # Reading course transcription and information
    elementss2 = driver.find_element_by_id('tab-overview').get_attribute('innerHTML')
    time.sleep(0.2)
    # Saving the information into the course directory as info.php file
    if not main.file_exist(Config.save_dir + r"\\" + cat_name + r"\\" + course_name + r"\info.php") or os.stat(Config.save_dir + r"\\" + cat_name + r"\\" + course_name + r"\info.php").st_size == 0:
       f = open(Config.save_dir + r"\\" + cat_name + r"\\" + course_name + r"\info.php", "a+", encoding="utf-8")
       f.write(elementss2)

    counter = 0       # Will count the loops - Just for the function
    temp_counter = 0  # Will count only the successfully downloaded videos for the current course
    time.sleep(0.2)

    # Getting the videos by Css selector
    # *Please note that it may differ after time if Lynda update their styles
    course = driver.find_elements_by_css_selector('.item-name.video-name.ga')

    # Will store all video links for the current course as an array
    listsak = []
    for k in course:
        # Adding the links
         listsak.append(k.get_attribute('href'))

    # Going trough every link in the array
    for i in listsak:
        counter += 1
        try:
             driver.get(i)   # Opening the video link     
        except:
            time.sleep(0.1)
            try:
                driver.get(i)   # Opening the video link   
            except:
                time.sleep(0.3)
                try:
                    driver.get(i)   # Opening the video link 
                except:
                         driver.get(i)   # Opening the video link 
        
        # Trying to click on play video button, if that did not work will look for modal close button and click on it
        try:
           WebDriverWait(driver, 100).until(EC.element_to_be_clickable((By.ID, "banner-play"))).click()
        except:
            # Trying to click on modal closing button
            driver.find_element_by_class_name("close").click()
            time.sleep(0.3)
            WebDriverWait(driver, 100).until(EC.element_to_be_clickable((By.ID, "banner-play"))).click()
            
        # 1 seconds waiting time works for my internet connection speed, but may not work with yours
        # If you see an error message url="" (an empty url), increase this sleeping time value
        time.sleep(1)

        # Trying to get the video tag and read the src attribute
        try:
            WebDriverWait(driver, 100).until(EC.element_to_be_clickable((By.TAG_NAME, "video")))
        finally:
            video = driver.find_element_by_tag_name('video')
        time.sleep(0.2)
        url_temp = video.get_attribute('src')

        # Clearing video name Stage 1
        video_name = driver.current_url.split("/")[4].replace("-", " ")
        # Clearing the download link to suits our needs
        dumper = video.get_attribute("data-conviva").replace("}", "").replace("{", "").replace("\"", "").replace(":", "-").split(",")
        # Adding numerated prefix for a proper lexicographic order
        video_name = "%04d_%s_%s_%s" % (counter, video_name, dumper[0].replace("Duration-", ""), dumper[4].replace("ReleaseYear-", ""))
        # Creating variable for the completed saving path
        save_dir   = Config.save_dir + r"\\" + cat_name + r"\\" + course_name + r"\\" + video_name + ".mp4"

        # Trying to download the video file three times
        # Each unsuccessful attempt will increase the sleeping time value
        # I have tested it and it works 100% for me with these delays
        # Keep an eye and increase or decrease the values if they do not work for you

        if not main.file_exist(save_dir):
            time.sleep(0.1)
            try:
                main.download(url_temp, save_dir)
            except:
                time.sleep(0.3)
                try:
                    main.download(url_temp, save_dir)
                except:
                    time.sleep(0.5)
                    try:
                        main.download(url_temp, save_dir)
                    except:
                        continue
            if Config.show_progress:
                 # Formatting and showing downloading procures bar
                 file_display = video_name.split("_")
                 main.show_progress_bar(url_temp, "\r"+file_display[1] + " " +
                                        str(datetime.timedelta(seconds=int(file_display[2]))), save_dir)
            else:
                # Trow a message if the current video file already exist
                sys.stdout.write("\r%s%s File %s successfully downloaded" % (
                    main.bcolors.sb, main.bcolors.fg, video_name + ".mp4"))
                sys.stdout.flush()

            total_counter += 1
            temp_counter += 1

        else:
            # Trow a message if the current video file already exist
            sys.stdout.write("\r%s%s File %s already exist. Skipped!" % (
                   main.bcolors.sb,main.bcolors.fr, video_name + ".mp4"))
            sys.stdout.flush()

        # Show message if all videos from the current course are downloaded
        if counter == len(course):
            courses_count += 1
            sys.stdout.write(
                  "\n%s[%s*%s]%s%sYou have successfully downloaded course %s%s %swith %d videos. Downloaded %d courses and %d videos in total" % (main.bcolors.bm, main.bcolors.fc, main.bcolors.fm, main.bcolors.fc,
                     main.bcolors.sd + main.bcolors.fc, main.bcolors.sb+main.bcolors.fc,course_name,main.bcolors.sd + main.bcolors.fc, temp_counter, courses_count, total_counter)
            )
            sys.stdout.flush()


